File: app/src/main/java/io/github/howiezuo/unsplash/feature/main/MainFragment.java
package io.github.howiezuo.unsplash.feature.main;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import io.github.howiezuo.unsplash.databinding.FragmentMainBinding;


public class MainFragment extends Fragment {

    private MainViewModel mainViewModel;

    private FragmentMainBinding mainBinding;

    public MainFragment() {
        //
    }

    public static MainFragment newInstance() {
        return new MainFragment();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mainBinding = FragmentMainBinding.inflate(inflater, container, false);
        mainBinding.setViewModel(mainViewModel);
        return mainBinding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        RecyclerView recyclerView = mainBinding.rvPhotos;
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new MainAdapter((MainListener) getActivity()));

        mainViewModel.create();
    }

    public void setViewModel(MainViewModel viewModel) {
        mainViewModel = viewModel;
    }
}

File: app/src/main/java/io/github/howiezuo/unsplash/feature/main/MainFragment.kt
package io.github.howiezuo.unsplash.feature.main

import android.databinding.Observable
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.github.howiezuo.unsplash.databinding.FragmentMainBinding
import io.github.howiezuo.unsplash.util.SnackbarUtils


class MainFragment : Fragment() {

    lateinit var mainViewModel: MainViewModel

    private lateinit var mainBinding: FragmentMainBinding

    private var snackbarCallback: Observable.OnPropertyChangedCallback? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mainBinding = FragmentMainBinding.inflate(inflater, container, false)
        mainBinding.viewModel = mainViewModel
        return mainBinding.root
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        snackbarCallback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable?, propertyId: Int) {
                SnackbarUtils.showSnackbar(getView(), mainViewModel.errorText.get())
            }
        }
        mainViewModel.errorText.addOnPropertyChangedCallback(snackbarCallback)

        mainBinding.srLayout.setOnRefreshListener {
            mainViewModel.refresh()
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        mainBinding.rvPhotos.layoutManager = LinearLayoutManager(context)
        mainBinding.rvPhotos.adapter = MainAdapter(activity as MainListener)

        mainViewModel.create()
    }

    override fun onDestroy() {
        if (snackbarCallback != null) {
            mainViewModel.errorText.removeOnPropertyChangedCallback(snackbarCallback)
        }
        super.onDestroy()
    }

    companion object {
        fun newInstance(): MainFragment {
            return MainFragment()
        }
    }

}
